src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs

```diff
@@ -67,61 +67,47 @@ protected override Task CleanupConnections()
+        protected override Task OnConnectedAsync(OpenConnectionMessage openConnectionMessage)
         {
             // Create empty transport with only channel for async processing messages
             var connectionId = openConnectionMessage.ConnectionId;
-            var clientContext = new ClientContext();
-            try
-            {
-                await clientContext.Output.WriteAsync(openConnectionMessage);
-                if(!_clientConnections.TryAdd(connectionId, clientContext))
-                {
-                    Log.DuplicateConnectionId(Logger, connectionId, null);
-                    throw new ArgumentException("ConnectionId already exists.");
-                }
+            var clientContext = new ClientContext(connectionId);
 
-                // Writing from the application to the service
-                _ = ProcessMessageAsync(connectionId);
+            if (_clientConnections.TryAdd(connectionId, clientContext))
+            {
+                _ = ProcessMessageAsync(clientContext);
+                return ForwardMessageToApplication(connectionId, openConnectionMessage);
             }
-            catch (Exception e)
+            else
             {
-                // Fail to write initial open connection message to channel
-                Log.ConnectedStartingFailed(Logger, connectionId, e);
-                // Close channel and notify client to close connection
-                clientContext.Output.TryComplete();
-                await WriteAsync(new CloseConnectionMessage(connectionId, e.Message));
+                Log.DuplicateConnectionId(Logger, connectionId, null);
+                return Task.CompletedTask;
             }
         }
 
-        protected override async Task OnDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
+        protected override Task OnDisconnectedAsync(CloseConnectionMessage closeConnectionMessage)
         {
-            var connectionId = closeConnectionMessage.ConnectionId;
-            if (_clientConnections.TryGetValue(connectionId, out var clientContext))
-            {
-                try
-                {
-                    await clientContext.Output.WriteAsync(closeConnectionMessage);
-                }
-                catch (Exception e)
-                {
-                    Log.FailToWriteMessageToApplication(Logger, connectionId, e);
-                }
-            }
+            return ForwardMessageToApplication(closeConnectionMessage.ConnectionId, closeConnectionMessage);
         }
 
-        protected override async Task OnMessageAsync(ConnectionDataMessage connectionDataMessage)
+        protected override Task OnMessageAsync(ConnectionDataMessage connectionDataMessage)
+        {
+            return ForwardMessageToApplication(connectionDataMessage.ConnectionId, connectionDataMessage);
+        }
+
+        private async Task ForwardMessageToApplication(string connectionId, ServiceMessage message)
```

Comment:
>ForwardMessageToApplication(string connectionId, ServiceMessage message) [](start = 27, length = 72)

Can simplify to ForwardMessageToApplication(ServiceMessage message) as ConnectionId is always get from ServiceMessage.

Reply:
However `ServiceMessage` class does not contain `ConnectionId` property, need a type check if we omit `connectionId`